### PR TITLE
Feature/14 verify 2 threshold

### PR DESCRIPTION
**Pull request by Rebal67**

### CODE REVIEW

File: src/JWT.ts [hunk 1]
```diff
@@ -250,16 +250,23 @@ export type GeneralJWSSignature = {
...
     const decodedJwt: JWTDecoded = Object.assign(jws, { payload: JSON.parse(decodeBase64url(jws.payload)) })
-    if (decodedJwt.header.cty === 'JWT') {
+    if (decodedJwt.header.cty === 'JWT' && recurse) {
+      signatures.push({
+        protected: decodedJwt.header,
+        signature: decodedJwt.signature as string,
+      })
       return decodeJWT(decodedJwt.payload.jwt)
     }
+    if (signatures.length != 0) decodedJwt.signature = signatures
```
Comment (theblockstalk) [inline src/JWT.ts]: !== does correct comparison

File: src/JWT.ts [hunk 2]
```diff
@@ -511,6 +519,15 @@ export async function verifyJWT(
         throw new Error(`${JWT_ERROR.INVALID_AUDIENCE}: JWT audience does not match your DID or callback url`)
       }
     }
+    if (header.cty === 'JWT') {
+      count++
+      return verifyJWT(payload.jwt, options, count)
```
Comment (theblockstalk) [inline src/JWT.ts]: this is nearly there but has a security problem. I can currently sign a JWT multiple times with the same key and get past the threshold. We need to verify that the signature condition matches from the required distinct and unique keys from the ConditionProof

File: src/VerifierAlgorithm.ts [hunk 3]
```diff
@@ -38,14 +39,16 @@ function extractPublicKeyBytes(pk: VerificationMethod): Uint8Array {
...
     // TODO: support multiple ThresholdCondition types and multiple public keys
-    return hexToBytes(
-      secp256k1
-        .keyFromPublic({
-          x: bytesToHex(base64ToBytes(pk.conditionWeightedThreshold[0].condition.publicKeyJwk.x)),
-          y: bytesToHex(base64ToBytes(pk.conditionWeightedThreshold[0].condition.publicKeyJwk.y)),
-        })
-        .getPublic('hex')
-    )
+    return pk.conditionWeightedThreshold.map((conditionWeight: ConditionWeightedThreshold) => {
```
Comment (theblockstalk) [inline src/VerifierAlgorithm.ts:42]: this should return extraPublicKeyBytes() again on the subtype of the condition.

see https://github.com/Tonomy-Foundation/did-resolver/pull/2#discussion_r1021278639

File: src/VerifierAlgorithm.ts [hunk 4]
```diff
@@ -155,14 +167,27 @@ export function verifyRecoverableES256K(
...
-      return (
-        keyHex === recoveredPublicKeyHex ||
-        keyHex === recoveredCompressedPublicKeyHex ||
-        pk.ethereumAddress?.toLowerCase() === recoveredAddress ||
-        pk.blockchainAccountId?.split('@eip155')?.[0].toLowerCase() === recoveredAddress || // CAIP-2
-        verifyBlockchainAccountId(recoveredPublicKeyHex, pk.blockchainAccountId) // CAIP-10
-      )
+      const publickeys = extractPublicKeyBytes(pk)
+      if (Array.isArray(publickeys)) {
+        return publickeys.some((publickey) => {
```
Comment (theblockstalk) [inline src/VerifierAlgorithm.ts:172]: some will check that 1 of M passes. We need to check that N (threshold) of M keys are verified

**Review by theblockstalk — CHANGES_REQUESTED** [review]
security issue needs fix. Other issues can be delt with in future work if desired.

File: src/VerifierAlgorithm.ts (hunk 4 above)
Comment (Rebal67) [reply src/VerifierAlgorithm.ts:172]: that gets checked in the did-jwt-vc

File: src/JWT.ts (hunk 2 above)
Comment (theblockstalk) [reply src/JWT.ts]: we decided this is too big to address this task and will be dealt with in follow up task https://github.com/Tonomy-Foundation/Antelope-SSI-Toolkit/issues/13

This must be fixed

**Pull request merged**